### DISCUSSION TITLE
Caddy + Mercure as a lambda function

### DISCUSSION THREAD

**Discussion by Houssem-Esprit** [brefphp/bref#999] [General]
Hi all, so I was playing with bref for a little time now and it was great so far, while doing so I had faced a use case where I needed to use caddy web server to deal with Mercure. Which is the best approach to implement this use case using bref?

**Comment by t-richard** [comment]
As far as I know, Mercure needs a persistent connection (ie. the client establishes a connection and keeps it open as long as it needs to) so that would most likely be incompatible with how API gateway and lambda works because of timeouts.

Also mercure is implemented in go so Bref won't really help here.

I think your best bet is probably to run the hub on ec2 or ECS/EKS. Or use the managed version provided by mercure.rocks

**Reply by Houssem-Esprit** [reply]
@t-richard so even bref supports docker won't help here?

**Reply by deleugpn** [reply]
I'm not familiar with Mercure, but if you need persistent connections it probably won't work. But if it is compatible with WebSocket Protocol, API Gateway does offer support for WebSockets in a way that AWS becomes responsible for keeping the connection open while Lambda is only triggered on a per-event basis with the provided payload.

**Reply by t-richard** [reply]
> @t-richard so even bref supports docker won't help here?

I don't think so. Again Mercure is a binary, written in Go so Bref won't help here. And using docker won't solve the persistent connection issue.

> But if it is compatible with WebSocket Protocol

Mercure is an alternative to WebSockets so you'll have to choose one or another. If you don't want to manage this infrastructure, I'd suggest either switch to WebSocket with API Gateway as Marco suggested or use the managed version at https://mercure.rocks/pricing

**Reply by t-richard** [reply]
FYI, because I don't trust my thin knowledge of mercure, I asked on the Symfony Slack the following question

> Hi, the docs state that
"The mercure protocol is [...] compatible with all existing servers, even those who don't support persistent connections (serverless architecture, PHP, FastCGI...)"
Is this talking about the server hosting the Hub, or is it only for the publishing part ?

And dunglas (the creator of Mercure) replied saying 
> it's for the publishing part
basically it means that as long as you have a hub (that **cannot be hosted on a lambda** or so), you just have to send a POST request to it and it works

So lambda is not an option to host the Hub.

Hope that helps !

**Reply by mnapoli** [reply]
FYI I've tested Websocket with API Gateway before, there's a bit of a setup to do with DynamoDB to keep connection IDs, but it worked well and it's really great to avoid having to maintain a server/maintain the persistent connection. Less risk for production too.

**Reply by Houssem-Esprit** [reply]
@t-richard thanks for all of these clarifications, the whole idea looks clear now. 
@mnapoli thanks, I think I will proceed with this choice.
